File: functions/src/models/Match.ts
/* eslint-disable */

import {TerrainAddress} from "./TerrainAddress";
import {BadModelFieldException} from "../exceptions/BadModelFieldException";
import {HasTerrainTypes} from "./constants/HasTerrainType";
import {SportConstants} from "./constants/SportConstants";
import {UserShortForm} from "./user/UserShortForm";
import {MatchDuration} from "./constants/MatchDuration";

export class Match {

    private id!: string;
    private sport!: string;
    private admin!: UserShortForm;
    private latitude: number = -1;
    private longitude: number = -1;
    private randomizedLocation: boolean = false;
    private inTemporalMatch!: boolean;
    private createdAtUTC!: number;
    private matchDetailsFromAdmin!: string;
    private matchDateInUTC!: number;
    private matchDuration!: number;
    private hasTerrainType!: string;
    private terrainAddress!: TerrainAddress | null;
    private levels!: number[];
    private userRequestsToJoinMatch!: string[];
    private adminIgnoredRequesters!: string[];
    private usersInChat!: string[];
    private lastModifiedTimeInUTC!: number;

    public static MIN_LEVEL:number = 1;
    public static MAX_LEVEL:number = 6;

constructor() {
    }

    static makeInstanceFromSnapshot(snapshotNeedsParsing: any):Match {

        const match:Match = new Match();

        let parsedSnapshot;
        try{
            parsedSnapshot = JSON.parse(snapshotNeedsParsing);
        }catch (Error){
            parsedSnapshot = snapshotNeedsParsing;
        }

        match.setId(parsedSnapshot.id);
        match.setSport(parsedSnapshot.sport);
        match.setAdmin(UserShortForm.makeInstanceFromSnapshot(parsedSnapshot.admin));
        match.setRandomizedLocation(parsedSnapshot.randomizedLocation);
        match.setLatitude(parsedSnapshot.latitude);
        match.setLongitude(parsedSnapshot.longitude);
        match.setInTemporalMatch(parsedSnapshot.inTemporalMatch);

        match.setCreatedAtUTC(parsedSnapshot.createdAtUTC);

        match.setMatchDetailsFromAdmin(parsedSnapshot.matchDetailsFromAdmin);
        match.setMatchDateInUTC(parsedSnapshot.matchDateInUTC);
        match.setMatchDuration(parsedSnapshot.matchDuration);
        match.setHasTerrainType(parsedSnapshot.hasTerrainType);


        let terrainAddress:TerrainAddress | null;
        if (parsedSnapshot.terrainAddress === undefined || parsedSnapshot.terrainAddress === null)
            terrainAddress = null;
        else
            terrainAddress = TerrainAddress.makeInstanceFromSnapshot(parsedSnapshot.terrainAddress);


        match.setTerrainAddress(terrainAddress)
        match.setLevels(parsedSnapshot.levels);
        match.setUserRequestsToJoinMatch(parsedSnapshot.userRequestsToJoinMatch);
        match.setAdminIgnoredRequesters(parsedSnapshot.adminIgnoredRequesters);
        match.setUsersInChat(parsedSnapshot.usersInChat);
        match.setLastModifiedTimeInUTC(parsedSnapshot.lastModifiedTimeInUTC);


        return match;
    }

    static makeCopy(matchToCopy: Match) {

        const match:Match = new Match();



        match.setId(matchToCopy.getId());
        match.setSport(matchToCopy.getSport());
        match.setAdmin(UserShortForm.makeCopy(matchToCopy.getAdmin()));
        match.setRandomizedLocation(matchToCopy.getRandomizedLocation());
        match.setLatitude(matchToCopy.getLatitude());
        match.setLongitude(matchToCopy.getLongitude());
        match.setInTemporalMatch(matchToCopy.getInTemporalMatch());

        match.setCreatedAtUTC(matchToCopy.getCreatedAtUTC());

        match.setMatchDetailsFromAdmin(matchToCopy.getMatchDetailsFromAdmin());
        match.setMatchDateInUTC(matchToCopy.getMatchDateInUTC());
        match.setMatchDuration(matchToCopy.getMatchDuration());
        match.setHasTerrainType(matchToCopy.getHasTerrainType());


        let terrainAddress:TerrainAddress | null;
        if (matchToCopy.getTerrainAddress() === undefined || matchToCopy.getTerrainAddress() === null)
            terrainAddress = null;
        else
            terrainAddress = TerrainAddress.makeCopy(matchToCopy.getTerrainAddress());


        match.setTerrainAddress(terrainAddress)
        match.setLevels(matchToCopy.getLevels());
        match.setUserRequestsToJoinMatch(matchToCopy.getUserRequestsToJoinMatch());
        match.setAdminIgnoredRequesters(matchToCopy.getAdminIgnoredRequesters());
        match.setUsersInChat(matchToCopy.getUsersInChat());
        match.setLastModifiedTimeInUTC(matchToCopy.getLastModifiedTimeInUTC());


        return match;
    }

    toObject() {//this is the output that goes to firestore so becareful with the changes in this class

        let terrainAddress:{} | null = null;
        if (this.terrainAddress !== null)
            terrainAddress = this.terrainAddress.toObject();


        return {
            id: this.id,
            sport: this.sport,
            admin: this.admin.toObject(),
            latitude: this.latitude,
            longitude: this.longitude,
            randomizedLocation: this.randomizedLocation,
            inTemporalMatch: this.inTemporalMatch,
            createdAtUTC: this.createdAtUTC,
            matchDetailsFromAdmin: this.matchDetailsFromAdmin,
            matchDateInUTC: this.matchDateInUTC,
            matchDuration: this.matchDuration,
            hasTerrainType: this.hasTerrainType,
            terrainAddress: terrainAddress,
            levels: this.levels,
            userRequestsToJoinMatch: this.userRequestsToJoinMatch,
            adminIgnoredRequesters: this.adminIgnoredRequesters,
            usersInChat: this.usersInChat,
            lastModifiedTimeInUTC: this.lastModifiedTimeInUTC,
            // add other fields as needed
        };
    }

    public getId(): string {
        return this.id;
    }

    public setId(id: string) {
        if (id === undefined || typeof id !== 'string') {
            throw new BadModelFieldException("id must be a string");
        }

        id = id.trim();
        if (id.length == 0)
            throw new BadModelFieldException("id length is 0");


        let maxLength:number = 60;
        if (id.length > maxLength){
            throw new BadModelFieldException("id must be less than " + maxLength + " letters");
        }

        this.id = id;
    }

    public getAdmin(): UserShortForm {
        return this.admin;
    }

    public setAdmin(admin: UserShortForm) {
        if (admin === undefined || !(admin instanceof UserShortForm)) {
            throw new BadModelFieldException("something is wrong with admin");
        }

        this.admin = admin;
    }

    public isAdmin(userId: string):boolean {

        return this.admin.getUserId() === userId;
    }


    public getLatitude(): number {
        return this.latitude;
    }

    public setLatitude(latitude: number) {
        if (latitude === undefined || typeof latitude !== 'number' || isNaN(latitude)) {
            throw new BadModelFieldException("latitude must be a valid number");
        }

        if (latitude < -90 || latitude > 90) {
            throw new BadModelFieldException("Invalid latitude. Latitude must be between -90 and 90 degrees.");
        }

        this.latitude = latitude;

        this.randomizeLocation(this.latitude,this.longitude);
    }

    public getLongitude(): number {
        return this.longitude;
    }

    public setLongitude(longitude: number) {
        if (longitude === undefined || typeof longitude !== 'number' || isNaN(longitude)) {
            throw new BadModelFieldException("longitude must be a valid number");
        }

        if (longitude < -180 || longitude > 180) {
            throw new BadModelFieldException("Invalid longitude. Longitude must be between -180 and 180 degrees.");
        }


        this.longitude = longitude;

        this.randomizeLocation(this.latitude,this.longitude);
    }

    private randomizeLocation(latitude: number, longitude: number): void {

        if (this.randomizedLocation)
            return;

        if (latitude === -1 || longitude === -1)
            return;

        // Maximum meters to change
        const maxDistanceMeters: number = 200;

        // Convert meters to degrees (approximately)
        const maxDegreeChange = maxDistanceMeters / 111000.0;

        // Generate random changes within the allowed range
        // Subtracting 0.5 to allow for both positive and negative changes
        const changeLat = (Math.random() - 0.5) * 2 * maxDegreeChange;
        const changeLon = (Math.random() - 0.5) * 2 * maxDegreeChange * Math.cos(latitude * Math.PI / 180);

        // Create new coordinates with these random changes
        const randomizedLatitude = latitude + changeLat;
        const randomizedLongitude = longitude + changeLon;

        this.randomizedLocation = true;

        this.latitude = randomizedLatitude;
        this.longitude = randomizedLongitude;
    }

    public getInTemporalMatch(): boolean {
        return this.inTemporalMatch;
    }

    public setInTemporalMatch(inTemporalMatch: boolean) {
        if (inTemporalMatch === undefined || typeof inTemporalMatch !== 'boolean') {
            throw new BadModelFieldException("inTemporalMatch must be a valid boolean");
        }

        this.inTemporalMatch = inTemporalMatch;
    }

    public getMatchDetailsFromAdmin(): string {
        return this.matchDetailsFromAdmin;
    }

    public setMatchDetailsFromAdmin(matchDetailsFromAdmin: string) {
        if (matchDetailsFromAdmin === undefined || matchDetailsFromAdmin === null){
            this.matchDetailsFromAdmin = "";
            return
        }

        if (matchDetailsFromAdmin === undefined || typeof matchDetailsFromAdmin !== 'string') {
            throw new BadModelFieldException("matchDetailsFromAdmin must be a string");
        }

        let matchDetailsFromAdminLength:number = 1000;
        if (matchDetailsFromAdmin.length > matchDetailsFromAdminLength){
            throw new BadModelFieldException("matchDetailsFromAdmin must be less than " + matchDetailsFromAdminLength + " letters");
        }

        this.matchDetailsFromAdmin = matchDetailsFromAdmin;
    }

    public getMatchDateInUTC(): number{
        return this.matchDateInUTC;
    }

    public setMatchDateInUTC(matchDateInUTC: number) {
        if (matchDateInUTC === undefined || typeof matchDateInUTC !== 'number' || isNaN(matchDateInUTC)) {
            throw new BadModelFieldException("createdAtUTC must be a valid number");
        }

        this.matchDateInUTC = matchDateInUTC;
    }

    public getMatchDuration(): number {
        return this.matchDuration;
    }

    public setMatchDuration(matchDuration: number) {
        if (matchDuration === undefined || typeof matchDuration !== 'number' || isNaN(matchDuration)) {
            throw new BadModelFieldException("match duration must be a number");
        }

        const minDuration:number = 0;
        if (matchDuration < minDuration)
            throw new BadModelFieldException("match duration must be higher than " + minDuration);


        if (!MatchDuration.checkDurationThatIsValidReturnBoolean(matchDuration))
            throw new BadModelFieldException("match duration is not valid");

        this.matchDuration = matchDuration;
    }

    public getHasTerrainType(): string{
        return this.hasTerrainType;
    }

    public setHasTerrainType(hasTerrainType: string) {
        if (hasTerrainType === undefined || typeof hasTerrainType !== 'string') {
            throw new BadModelFieldException("hasTerrainType must be a string");
        }

        let maxLength:number = 50;
        if (hasTerrainType.length > maxLength){
            throw new BadModelFieldException("has terrain type must be less than " + maxLength + " letters");
        }

        const found: boolean = HasTerrainTypes.TERRAIN_OPTIONS_ENGLISH_LIST.includes(hasTerrainType);
        if (!found)
            throw new BadModelFieldException("has terrain type bad value");

        this.hasTerrainType = hasTerrainType;
    }

    public getTerrainAddress(): TerrainAddress | null {
        return this.terrainAddress;
    }

    public setTerrainAddress(terrainAddress: TerrainAddress | null) {

        if (terrainAddress === null){
            this.terrainAddress = null;
            return
        }

        if (terrainAddress === undefined || !(terrainAddress instanceof TerrainAddress)) {
            throw new BadModelFieldException("something is wrong with terrainAddress");
        }


        this.terrainAddress = terrainAddress;
    }


    public getLevels(): number[] {
        return this.levels;
    }

    public setLevels(levels: number[]) {
        const minLevel = Match.MIN_LEVEL;
        const maxLevel = Match.MAX_LEVEL;
        for (const level of levels) {
            if (level === undefined || typeof level !== 'number' || isNaN(level) || !Number.isInteger(level)) {
                throw new BadModelFieldException("level must be a valid integer");
            }

            if (level < minLevel || level > maxLevel){
                throw new BadModelFieldException("level out of bounds");
            }

        }

        let deepCopy = JSON.parse(JSON.stringify(levels));


        this.levels = deepCopy;
    }

    public getUserRequestsToJoinMatch(): string[] {
        return this.userRequestsToJoinMatch;
    }

    public setUserRequestsToJoinMatch(userRequests: string[]) {

        for (const userRequest of userRequests) {
            if (userRequest === undefined || typeof userRequest !== 'string') {
                throw new BadModelFieldException("user request id must be a valid string");
            }

            let maxLength:number = 60;
            if (userRequest.length > maxLength){
                throw new BadModelFieldException("user request id must be less than " + maxLength + " letters");
            }
        }

        let deepCopy = JSON.parse(JSON.stringify(userRequests));


        this.userRequestsToJoinMatch = deepCopy;
    }

    public getAdminIgnoredRequesters(): string[] {

        return this.adminIgnoredRequesters;
    }

    public setAdminIgnoredRequesters(ignoredRequesters: string[]) {

        for (const ignored of ignoredRequesters) {
            if (ignored === undefined || typeof ignored !== 'string') {
                throw new BadModelFieldException("ignored id must be a valid string");
            }

            let maxLength:number = 60;
            if (ignored.length > maxLength){
                throw new BadModelFieldException("ignored must be less than " + maxLength + " letters");
            }
        }

        let deepCopy = JSON.parse(JSON.stringify(ignoredRequesters));

        this.adminIgnoredRequesters = deepCopy;
    }


    public isMember(userId:string): boolean {
        return this.getUsersInChat().includes(userId);
    }

    public getUsersInChat(): string[] {
        return this.usersInChat;
    }

    public setUsersInChat(usersInChat: string[]) {

        for (const user of usersInChat) {
            if (user === undefined || typeof user !== 'string') {
                throw new BadModelFieldException("user id must be a valid string");
            }

            let maxLength:number = 60;
            if (user.length > maxLength){
                throw new BadModelFieldException("user id must be less than " + maxLength + " letters");
            }
        }

        let deepCopy = JSON.parse(JSON.stringify(usersInChat));

        this.usersInChat = deepCopy;
    }

    public getLastModifiedTimeInUTC(): number {
        return this.lastModifiedTimeInUTC;
    }

    public setLastModifiedTimeInUTC(lastModifiedTimeInUTC: number) {
        const currentDate = new Date();

        this.lastModifiedTimeInUTC = currentDate.getTime();
    }


    public getSport(): string {
        return this.sport;
    }

    public setSport(sport: string) {

        if (sport === undefined || typeof sport !== 'string') {
            throw new BadModelFieldException("sport must be a valid string");
        }

        let maxLength:number = 50;
        if (sport.length > maxLength){
            throw new BadModelFieldException("sport must be less than " + maxLength + " letters");
        }

        const found: boolean = SportConstants.SPORTSLIST.includes(sport);
        if (!found)
            throw new BadModelFieldException("sport bad value");


        this.sport = sport;
    }

    public getCreatedAtUTC(): number {
        return this.createdAtUTC;
    }

    public setCreatedAtUTC(createdAtUTC: number) {

        if (createdAtUTC === undefined || typeof createdAtUTC !== 'number' || isNaN(createdAtUTC)) {
            throw new BadModelFieldException("created at must be a number");
        }

        this.createdAtUTC = createdAtUTC;
    }


    public isUserRequestedToJoin(requesterId: string): boolean {
        return this.userRequestsToJoinMatch.includes(requesterId);
    }

    private getRandomizedLocation(): boolean{
        return this.randomizedLocation
    }

    private setRandomizedLocation(randomizedLocation: boolean) {
        if (randomizedLocation === undefined || typeof randomizedLocation !== 'boolean') {
            throw new BadModelFieldException("matchSave must be a valid boolean");
        }

        this.randomizedLocation = randomizedLocation;
    }


}

File: app/src/main/java/com/baikas/sporthub6/models/Match.java
package com.baikas.sporthub6.models;

import androidx.annotation.NonNull;

import com.baikas.sporthub6.helpers.time.TimeFromInternet;
import com.baikas.sporthub6.interfaces.ObjectWithId;
import com.baikas.sporthub6.models.user.User;
import com.baikas.sporthub6.models.user.UserShortForm;
import com.firebase.geofire.GeoFireUtils;
import com.firebase.geofire.GeoLocation;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.Objects;

public class Match implements Serializable, ObjectWithId {

    @NotNull
    private String id;
    @NotNull
    private String sport;
    private UserShortForm admin;
    @NotNull
    private Double latitude;
    @NotNull
    private Double longitude;
    private boolean randomizedLocation = false;
    private boolean inTemporalMatch;
    private final long createdAtUTC;
    private String matchDetailsFromAdmin;
    private long matchDateInUTC;
    private long matchDuration;
    @NotNull
    private String hasTerrainType;
    @Nullable
    private TerrainAddress terrainAddress;
    @NotNull
    private List<Long> levels;
    @NotNull
    private List<String> userRequestsToJoinMatch;
    @NotNull
    private List<String> adminIgnoredRequesters;
    @NotNull
    private List<String> usersInChat;
    private long lastModifiedTimeInUTC;

    public static int MAX_LEVEL = 6;
    public static int MIN_LEVEL = 1;

    public Match(@NonNull String id){
        this.id = id;
        this.sport = "";
        this.matchDateInUTC = 99999999999999999L;
        this.createdAtUTC = 99999999999999999L;
    }

    public Match(@NonNull String id, @NotNull String sport, UserShortForm admin,
                 long createdAtUTC, String matchDetailsFromAdmin, boolean inTemporalMatch, long matchDateInUTC, long matchDuration, @NonNull String hasTerrainType, @Nullable TerrainAddress terrainAddress,
                 @NotNull List<Long> levels, @NotNull List<String> userRequestsToJoinMatch, @NotNull List<String> adminIgnoredRequesters, @NotNull List<String> usersInChat, @NonNull Double latitude, @NonNull Double longitude) {
        this.id = id;
        this.sport = sport;
        this.admin = admin;
        this.matchDetailsFromAdmin = matchDetailsFromAdmin;

        this.inTemporalMatch = inTemporalMatch;

        this.createdAtUTC = createdAtUTC;
        this.matchDateInUTC = matchDateInUTC;
        this.matchDuration = matchDuration;
        this.hasTerrainType = hasTerrainType;
        this.terrainAddress = terrainAddress;

        this.setLevels(levels);

        this.userRequestsToJoinMatch = userRequestsToJoinMatch;
        this.adminIgnoredRequesters = adminIgnoredRequesters;
        this.usersInChat = usersInChat;
        this.lastModifiedTimeInUTC = TimeFromInternet.getInternetTimeEpochUTC();

        this.randomizedLocation = false;
        this.setLatitude(latitude);
        this.setLongitude(longitude);
    }

    public Match(@NotNull Map<String,Object> map){
        this.id = (String) map.get("id");
        this.sport = (String) map.get("sport");
        this.admin = new UserShortForm((Map<String, Object>) map.get("admin"));

        if (map.get("createdAtUTC") instanceof Integer){
            this.createdAtUTC = ((Integer) map.get("createdAtUTC")).longValue();
        }else
            this.createdAtUTC = (Long) map.get("createdAtUTC");


        this.matchDetailsFromAdmin = (String) map.get("matchDetailsFromAdmin");

        if (map.get("matchDateInUTC") instanceof Integer){
            this.matchDateInUTC = ((Integer) map.get("matchDateInUTC")).longValue();
        }else
            this.matchDateInUTC = (Long) map.get("matchDateInUTC");

        if (map.get("matchDuration") instanceof Integer){
            this.matchDuration = ((Integer) map.get("matchDuration")).longValue();
        }else
            this.matchDuration = (Long) map.get("matchDuration");

        this.hasTerrainType = (String) map.get("hasTerrainType");

        if (map.get("terrainAddress") == null)
            this.terrainAddress = null;
        else
            this.terrainAddress = new TerrainAddress((Map<String,Object>) map.get("terrainAddress"));

        this.lastModifiedTimeInUTC = TimeFromInternet.getInternetTimeEpochUTC();

        this.inTemporalMatch = (Boolean) map.get("inTemporalMatch");

        this.setLevels((List<Long>) map.get("levels"));

        this.userRequestsToJoinMatch = (List<String>) map.get("userRequestsToJoinMatch");
        this.adminIgnoredRequesters = (List<String>) map.get("adminIgnoredRequesters");
        this.usersInChat = (List<String>) map.get("usersInChat");

        this.randomizedLocation = (Boolean)map.get("randomizedLocation");
        this.setLatitude((Double) map.get("latitude"));
        this.setLongitude((Double) map.get("longitude"));

    }

    public Match(@NotNull Match match){
        this.id = match.id;
        this.sport = match.sport;

        if (match.admin == null)
            this.admin = null;
        else
            this.admin = new UserShortForm(match.admin);


        this.createdAtUTC = match.createdAtUTC;
        this.matchDetailsFromAdmin = match.matchDetailsFromAdmin;

        this.inTemporalMatch = match.inTemporalMatch;

        this.matchDateInUTC = match.matchDateInUTC;
        this.matchDuration = match.matchDuration;
        this.hasTerrainType = match.hasTerrainType;

        if (match.terrainAddress == null)
            this.terrainAddress = null;
        else
            this.terrainAddress = new TerrainAddress(match.terrainAddress);

        this.lastModifiedTimeInUTC = TimeFromInternet.getInternetTimeEpochUTC();

        this.levels = new ArrayList<>(match.levels); // Create a new List to avoid sharing the reference
        this.userRequestsToJoinMatch = new ArrayList<>(match.userRequestsToJoinMatch); // Create a new List
        this.adminIgnoredRequesters = new ArrayList<>(match.adminIgnoredRequesters); // Create a new List
        this.usersInChat = new ArrayList<>(match.usersInChat);

        this.randomizedLocation = match.isRandomizedLocation();
        this.latitude = match.getLatitude();
        this.longitude = match.getLongitude();
    }



    @Override
    public boolean equals(Object o) {//we dont use lastUpdateTime
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        Match match = (Match) o;
        return createdAtUTC == match.createdAtUTC && matchDateInUTC == match.matchDateInUTC && inTemporalMatch == match.inTemporalMatch && Objects.equals(admin,match.getAdmin()) && Objects.equals(matchDuration,match.matchDuration) && Objects.equals(matchDetailsFromAdmin,match.matchDetailsFromAdmin) && hasTerrainType.equals(match.hasTerrainType) && Objects.equals(terrainAddress, match.terrainAddress) && id.equals(match.id) && sport.equals(match.sport) && latitude.equals(match.latitude) && longitude.equals(match.longitude) && randomizedLocation == match.randomizedLocation && levels.equals(match.levels) && userRequestsToJoinMatch.equals(match.userRequestsToJoinMatch) && adminIgnoredRequesters.equals(match.adminIgnoredRequesters) && usersInChat.equals(match.usersInChat);
    }

    @Override
    public int hashCode() {//we dont use admin (User) and lastUpdateTime and matchSave
        return Objects.hash(id, sport, latitude, longitude, randomizedLocation,createdAtUTC, matchDateInUTC, inTemporalMatch, admin, matchDuration, hasTerrainType, matchDetailsFromAdmin,terrainAddress, levels, userRequestsToJoinMatch, adminIgnoredRequesters,usersInChat);
    }

    @NonNull
    public String getId() {
        return id;
    }


    @NonNull
    public List<String> getUserRequestsToJoinMatch() {
        return userRequestsToJoinMatch;
    }

    public boolean isUserRequestedToJoin(String userId) {
        return userRequestsToJoinMatch.contains(userId);
    }

    public void setId(@NotNull String id) {
        this.id = id;
    }

    public void setUserRequestsToJoinMatch(@NonNull List<String> userRequestsToJoinMatch) {
        this.userRequestsToJoinMatch = userRequestsToJoinMatch;
    }

    public void setSport(String sport) {
        this.sport = sport;
    }

    @NonNull
    public String getSport() {
        return sport;
    }


    public long getCreatedAtUTC() {
        return createdAtUTC;
    }

    @NotNull
    public String getHasTerrainType() {
        return hasTerrainType;
    }

    public void setHasTerrainType(@NotNull String hasTerrainType) {
        this.hasTerrainType = hasTerrainType;
    }

    @Nullable
    public TerrainAddress getTerrainAddress() {
        return terrainAddress;
    }

    public void setTerrainAddress(@Nullable TerrainAddress terrainAddress) {
        this.terrainAddress = terrainAddress;
    }

    @NonNull
    public Double getLatitude() {
        return latitude;
    }

    public void setLatitude(@NotNull Double latitude) {
        this.latitude = latitude;
    }

    public boolean isRandomizedLocation() {
        return randomizedLocation;
    }

    public void setRandomizedLocation(boolean randomizedLocation) {
        this.randomizedLocation = randomizedLocation;
    }

    public String getMatchDetailsFromAdmin() {
        return matchDetailsFromAdmin;
    }

    public void setMatchDetailsFromAdmin(String matchDetailsFromAdmin) {
        this.matchDetailsFromAdmin = matchDetailsFromAdmin;
    }

    @NonNull
    public Double getLongitude() {
        return longitude;
    }

    public void setLongitude(@NotNull Double longitude) {
        this.longitude = longitude;

    }



    public long getMatchDuration() {
        return matchDuration;
    }

    public void setMatchDuration(long matchDuration) {
        this.matchDuration = matchDuration;
    }

    @NotNull
    public List<String> getUsersInChat() {
        return usersInChat;
    }

    public void setUsersInChat(@NotNull List<String> usersInChat) {
        this.usersInChat = usersInChat;
    }

    public boolean isMember(String userId){
        return this.usersInChat.contains(userId);
    }

    public long getMatchDateInUTC() {
        return matchDateInUTC;
    }

    public void setMatchDateInUTC(long matchDateInUTC) {
        this.matchDateInUTC = matchDateInUTC;
    }

    @NotNull
    public List<String> getAdminIgnoredRequesters() {
        return adminIgnoredRequesters;
    }

    public void setAdminIgnoredRequesters(@NotNull List<String> adminIgnoredRequesters) {
        this.adminIgnoredRequesters = adminIgnoredRequesters;
    }


    @NonNull
    public List<Long> getLevels() {
        return levels;
    }

    public void setLevels(@NonNull List<Long> levels) {
        for (int i = 0; i < levels.size(); i++) {
            Number number = levels.get(i);
            if (number instanceof Integer){
                long makeItLong = (number).longValue();
                levels.set(i, makeItLong);
            }

        }
        this.levels = levels;
    }

    public long getLastModifiedTimeInUTC() {
        return lastModifiedTimeInUTC;
    }

    public void setLastModifiedTimeInUTC(long lastModifiedTimeInUTC) {
        this.lastModifiedTimeInUTC = lastModifiedTimeInUTC;
    }

    public boolean isInTemporalMatch() {
        return inTemporalMatch;
    }

    public void setInTemporalMatch(boolean inTemporalMatch) {
        this.inTemporalMatch = inTemporalMatch;
    }

    public UserShortForm getAdmin() {
        return admin;
    }

    public void setAdmin(UserShortForm admin) {
        this.admin = admin;
    }

    public boolean isAdmin(String adminId) {
        return this.admin.getUserId().equals(adminId);
    }


}
